feat(rewrite): document rewrite APIs (#8914)
withastro/docs: src/content/docs/en/guides/routing.mdx

</html>
```

## Rewrites

<Since v="4.13.0" />

> A rewrite allows you serve a different route without redirecting the browser to a different page. The browser will show the same address in the URL bar, but will display the content of the URL provided to `Astro.rewrite()`.

Reviewer: ```suggestion
A rewrite allows you serve a different route without redirecting the browser to a different page. The browser will show the same address in the URL bar, but will display the content of the URL provided to [`Astro.rewrite()`](/en/reference/api-reference/astrorewrite).
```
Realized we should have a link to the reference here!

> This can be useful for showing the same content at multiple paths (e.g. `/products/shoes/men/` and `/products/men/shoes`/).

Reviewer: ```suggestion
Rewrites can be useful for showing the same content at multiple paths (e.g. `/products/shoes/men/` and `/products/men/shoes`/).
```

For content that has permanently moved, or to direct your user to a different page (e.g. a user dashboard after logging in), use a [redirect](#redirects) instead.

> A rewrite is really useful in an i18n context, for example. Let's suppose your website serves localised version of Spanish e.g. `/es-es/articles/introduction` (Spain) and `/es-mx/articles/introduction` (Mexico). However, your website also serves pages for `es-CU` (Cuba), but you're missing the localised content for `/es-cu/articles/introduction`. 
> 
> However, you don't to serve that page as a 404, and you don't want to serve a redirect for SEO purpose. That's where the rewrites comes into play. You can decide to **fallback** to the Mexican localised page using the new `Astro.redirect` method:

Reviewer: ```suggestion
Rewrites are also useful to provide fallback content, such as for a localized page route whose content does not exist. If you don't want to serve a redirect for SEO purposes, and do not want to show a 404, then you can use a rewrite to display a different page at the requested route. The following example uses a rewrite to display the `/es-mx/` version of a page (Mexican Spanish) without from the `/es-CU/` (Cuban Spanish) URL path. 
```

Just rough draft for now, but I'm thinking all of this could be condensed to something like this...

```astro title="src/pages/es-cu/articles/introduction.astro"
---

> return Astro.rewrite("/es-mx/articles/introduction")

Reviewer: ```suggestion
return Astro.rewrite("/es/articles/introduction")
```
See the comment I made above for the corresponding slight change in the example -- I think it's a simpler example that someone wants to just write one general Spanish page and use that for variants? (At least, is visually simpler, so the idea gets across better, I think!)

Author: I leave the example to you, since you have more hands-on experience :)

---
```

> When you'll navigate the URL `/es-cu/articles/introduction`, Astro will render the route `src/pages/es-mx/articles/introduction.astro`.

Reviewer: ```suggestion
In the previous example, when a visitor navigates to the URL `/es-cu/articles/introduction`, Astro will render the content generated by the file `src/pages/es/articles/introduction.astro`
```

> The function `Astro.redirect()` will trigger a new complete rendering phase, which means that, if you have middleware, it will be executed again for the new route/request. 

Reviewer: This feels like maybe it should go somewhere before the example, as it's more general about how it works. It also feels like it might get lost or grouped in with the example, when this is a standalone fact. I'll try moving it up and see what we think! 

```suggestion
```

</html>
```

## Rewrites

<p><Since v="4.13.0" /></p>

> A rewrite allows you serve a different route without redirecting the browser to a different page. The browser will show the original address in the URL bar, but will instead display the content of the URL provided to [`Astro.rewrite()`](/en/reference/api-reference/astrorewrite).

Reviewer: ```suggestion
A rewrite allows you serve a different route without redirecting the browser to a different page. The browser will show the original address in the URL bar, but will instead display the content of the URL provided to [`Astro.rewrite()`](/en/reference/api-reference/#astrorewrite).
```

> For content that has permanently moved, or to direct your user to a different page (e.g. a user dashboard after logging in), use a [redirect](#redirects) instead.

Reviewer: ```suggestion
:::tip
For content that has permanently moved, or to direct your user to a different page with a new URL (e.g. a user dashboard after logging in), use a [redirect](#redirects) instead.
:::
```
Feels tippy to me!

</html>
```

## Rewrites

<p><Since v="4.13.0" /></p>

A rewrite allows you serve a different route without redirecting the browser to a different page. The browser will show the original address in the URL bar, but will instead display the content of the URL provided to [`Astro.rewrite()`](/en/reference/api-reference/#astrorewrite).

:::tip
For content that has permanently moved, or to direct your user to a different page with a new URL (e.g. a user dashboard after logging in), use a [redirect](#redirects) instead.
:::

Rewrites can be useful for showing the same content at multiple paths (e.g. `/products/shoes/men/` and `/products/men/shoes`/) without needing to maintain two different source files.

> Rewrites are also useful for SEO purposes and user experience. They allow you to display content that otherwise would require redirecting your visitor to a different page or would take them to a 404 page. One common use of rewrites is to show the same localized content for different variants of a language.

Reviewer: ```suggestion
Rewrites are also useful for SEO purposes and user experience. They allow you to display content that otherwise would require redirecting your visitor to a different page or would return a 404 status. One common use of rewrites is to show the same localized content for different variants of a language.
```

The following example uses a rewrite to render the `/es/` version of a page when the `/es-CU/` (Cuban Spanish) URL path is visited. When a visitor navigates to the URL `/es-cu/articles/introduction`, Astro will render the content generated by the file `src/pages/es/articles/introduction.astro`.

```astro title="src/pages/es-cu/articles/introduction.astro"
---
return Astro.rewrite("/es/articles/introduction")
---
```

> 

Reviewer: ``````suggestion

Use context.rewrite in your endpoint files to reroute to a different page:

```js
export function GET(ctx) {
  if (!ctx.locals.allowed) {
    return ctx.rewrite("/")
  }
}
```

``````
We didn't actually show an endpoint rewrite here, so we probably should!

:::

Rewrites can be useful for showing the same content at multiple paths (e.g. `/products/shoes/men/` and `/products/men/shoes`/) without needing to maintain two different source files.

Rewrites are also useful for SEO purposes and user experience. They allow you to display content that otherwise would require redirecting your visitor to a different page or would return a 404 status. One common use of rewrites is to show the same localized content for different variants of a language.

The following example uses a rewrite to render the `/es/` version of a page when the `/es-CU/` (Cuban Spanish) URL path is visited. When a visitor navigates to the URL `/es-cu/articles/introduction`, Astro will render the content generated by the file `src/pages/es/articles/introduction.astro`.

```astro title="src/pages/es-cu/articles/introduction.astro"
---
return Astro.rewrite("/es/articles/introduction")
---
```

Use `context.rewrite()` in your endpoint files to reroute to a different page:

```js title="src/pages/api.js"

> export function GET(ctx) {
>   if (!ctx.locals.allowed) {
>     return ctx.rewrite("/")

Reviewer: ```suggestion
export function GET(context) {
  if (!context.locals.allowed) {
    return context.rewrite("/")
```

  }
}
```

If the URL passed to `Astro.rewrite()` emits a runtime error, Astro will show the overlay error in development and return a 500 status code in production. If the URL does not exist in your project, a 404 status code will be returned.

You can intentionally create a rewrite to render your `/404` page, for example to indicate that a product in your e-commerce shop is no longer available:

```astro title="src/pages/[item].astro"
---
const { item } = Astro.params;

if (!itemExists(item)) {
  return Astro.rewrite("/404")
}
---

<div>...</div>
``` 

You can also conditionally rewrite based on an HTTP response status, for example to display a certain page on your site when visiting a URL that doesn't exist:

```js title="src/middleware.mjs"

> export const onRequest = async (ctx, next) => {
>   const response = await next();
>   if (response.status === 404) {
>     return ctx.rewrite("/");

Reviewer: ```suggestion
export const onRequest = async (context, next) => {
  const response = await next();
  if (response.status === 404) {
    return context.rewrite("/");
```
